pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,306 @@
+	mcClient mcclientset.Interface,
+	podInformer cache.SharedIndexInformer,
+	gwInformer cache.SharedIndexInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCPodRouteController {
+	controller := &MCPodRouteController{
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteController"),
+		podInformer:     podInformer,
+		podLister:       corelisters.NewPodLister(podInformer.GetIndexer()),
+		gwInformer:      gwInformer,
+		gwLister:        mclisters.NewGatewayLister(gwInformer.GetIndexer()),
+		namespace:       namespace,
+		installedPods:   make(map[string]string),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+
+	controller.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *MCPodRouteController) enqueueGateway(obj interface{}, isDelete bool) {
+	gw, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueueGateway can't process event", "obj", obj)
+		return
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	// Gateway webhook will guarantee that at most one Gateway in a cluster.
+	if isDelete {
+		c.activeGWName = ""
+		if c.nodeConfig.Name != gw.Name {
+			klog.InfoS("This Node is not the deleted Gateway, skip it", "node", c.nodeConfig.Name, "gateway", gw.Name)
+			return
+		}
+
+		for podIP := range c.installedPods {
+			c.queue.Add(&endpointChangeInfo{
+				podIP: podIP,
+			})
+			delete(c.installedPods, podIP)
+		}
+		return
+	}
+
+	c.activeGWName = gw.Name
+	if c.nodeConfig.Name != gw.Name {
+		klog.InfoS("The Node is not the Gateway, skip it", "node", c.nodeConfig.Name, "gateway", gw.Name)
+		return
+	}
+	pods, _ := c.podLister.List(labels.Everything())
+	for _, p := range pods {
+		pod := p
+		isValidPod := c.validatePod(pod)
+		if !isValidPod {
+			continue
+		}
+		c.queue.Add(&endpointChangeInfo{
+			podIP:     pod.Status.PodIP,
+			podNodeIP: pod.Status.HostIP,
+		})
+		c.installedPods[pod.Status.PodIP] = pod.Status.HostIP
```

Comment:
Probably rename `installedPods` as these are not really installed Pods. Something like `podNodeIPs`.